gdc_ng_models/models/misc.py

```diff
@@ -22,7 +23,7 @@ class FileReport(Base):
     node_id = Column('node_id', Text, index=True)
     ip = Column('ip', String)
     country_code = Column('country_code', String, index=True)
-    timestamp = Column('timestamp', DateTime, server_default="now()")
+    timestamp = Column('timestamp', DateTime, server_default=func.now(), primary_key=True)
```

Comment:
Creating an index on `timestamp` makes sense since we want to partition on this column.  Changing the primary key to be a composite key of `id` and `timestamp` seems off to me.  Is there a reason why we need to do this? A new index on `timestamp` seems better but I haven't used partition in pg.  So, I'm unfamiliar with the gotchas that come along with  it.

Btw, multi-column indexes in pg care about the order of columns.  This composite primary index will be id, then timestamp.  So, we only the benefits of the index when querying timestamp if we're also querying id.  But since we're partitioning only timestamp, that makes me wonder if this will work as intended.  (Which is another reason why a new index on timestamp might be preferred).

Reply:
The reason we have to add timestamp to the primary key constraint is because of the limitation when partitioning the table:
[REF](https://www.postgresql.org/docs/current/ddl-partitioning.html)
`Unique constraints (and hence primary keys) on partitioned tables must include all the partition key columns. This limitation exists because PostgreSQL can only enforce uniqueness in each partition individually.`

The point about order being important when creating multi-column indexes is valid. In this case, a query with `timestamp` only will not use the index. We can create another index on timestamp only. I'm not sure if the column is more read heavy or update heavy. An additional index might impact INSERT or UPDATE performance.

Reply:
Almost all (99.99%) of the activity looks like this:

1. Insert a record when a download starts.
2. Read the record back when the download ends.
3. Update the record with information computed at the end of the download.

Beyond that, we run a monthly report to aggregate the past month's data, and we very occasionally run ad-hoc queries, but performance isn't especially important for these aggregations as long as they run reliably.

Because of the end-of-download lookup, it is important that the API can retrieve by ID efficiently. However, we may want to add the timestamp to the [lookup](https://github.com/NCI-GDC/gdcapi/blob/acc2707a41733f3ba5fffbe86fad523f2b406d27/gdcapi/download/report.py#L87) anyway so that Postgres knows which partition to check.

My off-the-cuff guess would be not to do anything with the column order or indexes because I don't think a timestamp-only index would greatly benefit our query pattern and I don't want the current form of the API's update queries (without the timestamp) to become extremely slow. If we assume the API is updated, having the timestamp come first might make more sense, but I don't know if we want to assume that. Performance testing in QA might tell us something.

Reply:
ah. i didn't know this was a limitation of pg partitioning.  so, we *must* use a composite key. this is an unfortunate trade off, as we can now have duplicate ids as long as the timestamp differs. but the benefits of partitioning outweigh this unlikely event.  (it'd require a manual insert or a serious bug with sequences in pg) 

i learned something new today!

re: composite index ordering:  if we find that partitioning by timestamp is slow, we could add a second unique index that swaps the order.  it's feels dirty.  but it'd allow both id-only and timestamp-only lookups to happen fast.  or we could update gdcapi to query with both id and timestamp, as greg alluded to.  either way, i agree that we should do a load test first to see if this even matters.